walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,133 @@
+      if (!attendees.isEmpty()) {
+        for (String attendee : attendees) {
+          if (event.getAttendees().contains(attendee) && !eventsForAttendees.contains(event)) {
+            eventsForAttendees.add(event);
+          }
+        }
+      }
+    }
+
+    return eventsForAttendees;
+  }
+
+  public Collection<TimeRange> getPossibleTimes(List<Event> eventsWithAttendees, MeetingRequest request) {
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    for (int counter = 0; counter < eventsWithAttendees.size(); counter++) {
+      Event currentEvent = eventsWithAttendees.get(counter);
+
+      // If the array is not empty, then it means there was one event before this one. 
+      if (!availableTimes.isEmpty()) {
+        Event previousEvent = eventsWithAttendees.get(counter - 1);
+
+        // Remove the last added timerange to consider the next event. 
+        availableTimes.remove(availableTimes.size() - 1);
+
+        // If one event fully contains another (nested event), there are 2 possible times: before and after the longest event. 
+        if (previousEvent.getWhen().end() > currentEvent.getWhen().start()
+            && previousEvent.getWhen().end() > currentEvent.getWhen().end()) {
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+          break;
+        }
+
+        // If events have no conflict, there are 3 possible times: before the first one, between the 2 events and after the second one.
+        long durationBetweenEvents = currentEvent.getWhen().start() - previousEvent.getWhen().end(); 
+        if (request.getDuration() <= durationBetweenEvents) 
+          availableTimes.add(TimeRange.fromStartEnd(previousEvent.getWhen().end(), currentEvent.getWhen().start(), false));
+
+        // If the end of the event is not the end of the day, then the time in between is a possible time. 
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1) {
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        }
+      }
+
+      // If there were no previous times, then this event is the first one of the day. The possible times are before and after the event.
+      else {
+        // If the first event of the day starts at the beginning of the day, there are no availabilities before it. 
+        if (currentEvent.getWhen().start() != TimeRange.START_OF_DAY) 
+          availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.getWhen().start(), false));
+  
+        if (currentEvent.getWhen().end() != TimeRange.END_OF_DAY + 1)
+          availableTimes.add(TimeRange.fromStartEnd(currentEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    return availableTimes;
+  }
+
+  //Return the events common to both required and optional attendees
+  public Collection<TimeRange> getCommonTimes(List<Event> eventsWithRequiredAttendees,
```

Comment:
Is this special case (merging time from optional and required attendees) necessary?

I'd have thought you could just call 'getEventsForAttendees' with the combined list of attendees, and call 'getPossibleTimes' once, but I might be missing something.

Reply:
This is the reasoning behind required and optional attendees: https://github.com/helenatran/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md#writing-more-tests

Here, I tried to consider the scenarios where we would ignore the optional attendees availabilities (e.g.: when they are not available, or the optional attendees availabilities result in smaller time slots than required, etc.). Let me know whether some of my code might be redundant or unclear to achieve this.

Reply:
Thanks, is that the 'if(commonAvailabilities.isEmpty())' code near the bottom? I guess you could have the same logic with slightly simpler code by calling getPossibleTimes once with the combined set of attendees, then if that result is empty calling it with only the required attendees? It's definitely fine either way though